### PR TITLE
Play computer

### PR DESCRIPTION
**Pull request by Fiphee**
Play versus computer with 3 selectable difficulties

### CODE REVIEW

File: game/board.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def show(board):
 
 
 def set_choice(board, choice, sign):
-    new_board = board.copy()
+    new_board = [x[:] for x in board]
```
Comment (mihaitzzza) [inline game/board.py:53]: Please use `copy.deepcopy()` discuss yesterday, it's a bit cleaner and easier to read.

File: game/menus.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+from .modes import play
+
+def main_menu():
```
Comment (mihaitzzza) [inline game/menus.py:3]: I suppose here's a warning for two empty lines expected :)

File: game/menus.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    '''
+    Displays main menu options
+    '''                                       
+    while True:
```
Comment (mihaitzzza) [inline game/menus.py:7]: `while selection != '2'`

File: game/menus.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        print('\nMain Menu')
+        print('----------')
+        print('1. Play')  # play submenu function created below
+        print('2. Exit')  # add more print lines for other submenus
+        
+        selection = input('>: ')
+        if selection == '1':  # add an if check for the new menu added
+            print('Selected Play')
+            play_menu()  # call menu or feature function
+        elif selection == '2':
```
Comment (mihaitzzza) [inline game/menus.py:17]: Not needed `else` branch and manual `break` if using the right condition in the `while` loop.

File: game/menus.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            break
+
+def play_menu():
+    '''
+    Displays play menu options
+    '''
+    on_play_menu = True   
+    while on_play_menu:
```
Comment (mihaitzzza) [inline game/menus.py:25]: `while selection != '3'`

File: game/menus.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        print('\nPlay')
+        print('----------')
+        print('1. Vs Player')
+        print('2. Vs Computer')
+        print('3. Return to Main Menu')
+        
+        selection = input('>: ')
+        if selection == '1':
+            play()  
+            on_play_menu = False
```
Comment (mihaitzzza) [inline game/menus.py:35]: I would add a `break` here instead of using the `on_play_menu` variable.

File: game/menus.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+        elif selection == '2':
+            select_difficulty_menu()
+            on_play_menu = False
```
Comment (mihaitzzza) [inline game/menus.py:39]: I would add a `break` here instead of using the `on_play_menu` variable.

File: game/menus.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+        elif selection == '3':
```
Comment (mihaitzzza) [inline game/menus.py:41]: Remove these lines by simply change the condition on `while` loop.

File: game/menus.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            on_play_menu = False
+
+def select_difficulty_menu():
+    '''
+    Displays CPU difficulties to choose from
+    '''
+    on_difficulty_menu = True
+    while on_difficulty_menu:
```
Comment (mihaitzzza) [inline game/menus.py:49]: Same as the `while` comments from `play_menu` function.

File: game/modes.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            logger.error('Your choice is not an option.')
+            logger.exception(e)
+            is_correct_choice = False
+            continue
+        else:
+            is_correct_choice = True
+            logger.info('Player choice: %s' % choice)
+            return is_correct_choice, choice
+
+def play(cpu=False):
```
Comment (mihaitzzza) [inline game/modes.py:103]: Please use `None` as default. You're sending a `string` from the `menu` functions, but you're defaulting to a `boolean`. I know Python allow you to change the type of a variable, but try to avoid it.

File: game/modes.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+                if choice % 2 == 0:
+                    return choice
+    if last_move % 2 == 0:
+        if 5 in available_options:
+            return 5
+
+    return random.choice(corners)
+
+def _player_move(board_matrix, available_options):
+    while True:
```
Comment (mihaitzzza) [inline game/modes.py:81]: `while not is_correct_choice`

File: game/modes.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        except ValueError as e:
+            print('Your choice is not an option.')
+            logger.error('Your choice is not an option.')
+            logger.exception(e)
+            is_correct_choice = False
+            continue
+        else:
+            is_correct_choice = True
+            logger.info('Player choice: %s' % choice)
+            return is_correct_choice, choice
```
Comment (mihaitzzza) [inline game/modes.py:101]: Using the `while` condition from my comment, you can get rid of the `else` branch and do the return outside the `while` loop - code simplification.

File: game/modes.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        show(board_matrix)
+        print('Pick a choice from available options: ', available_options)
+        logger.info('Pick a choice from available options: %s' % available_options)
+
+        choice = input('Pick your choice: ')
+        logger.info('Pick your choice: ')
+        try:
+            choice = int(choice)
+            if choice not in available_options:
+                raise ValueError()
```
Comment (mihaitzzza) [inline game/modes.py:91]: Add a message to explain the error. `Not an available choice` or something.

File: game/modes.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    if there is no winning move next move, it chooses a random available option
+    'hard' checks for win or blocks player like the medium difficulty, but also strategically picks a move
+    if there is no winning or blocking move
+    
+    return:: choice
+    '''
+    if difficulty == 'easy':
+        choice = random.choice(available_options)
+        return choice
+
```
Comment (mihaitzzza) [inline game/modes.py:46]: Remove not needed extra line.

File: game/modes.py [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    elif difficulty == 'medium':
```
Comment (mihaitzzza) [inline game/modes.py:47]: What about the `'hard'` one? :) What happens if I select the HARD difficulty?

File: game/modes.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+def _found_winning_cell(board_matrix, available_options):
+    for choice in available_options:
+        check_spots = [x[:] for x in board_matrix]
+        new_board = set_choice(check_spots, choice, 'o')
+        is_won, _ = check_status(new_board)
+        if is_won:
+            return True, choice
+    return False, None
+
+def _found_blocking_cell(board_matrix, available_options):
```
Comment (mihaitzzza) [inline game/modes.py:19]: `_find_blocking_cell` or `_get_blocking_cell`. Don't use `found`. The functions/methods should express actions (present time).
Also just realize this function duplicates 99% of the previous one. The only difference is the `'x'` or `'o'` parameter sent to `set_choice`. Please refactor the previous function to allow another parameter and don't duplicate the code.

File: game/modes.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
+from .board import board_matrix as initial_board, get_options, show, set_choice
+from .status import check_status
+from .player import get_current_player
+import logging
+import random
+
+logger = logging.getLogger(__name__)
+
+
+def _found_winning_cell(board_matrix, available_options):
```
Comment (mihaitzzza) [inline game/modes.py:10]: `_find_winning_cell` or `_get_winning_cell`. Don't use `found`. The functions/methods should express actions (present time).

File: game/modes.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    for choice in available_options:
+        check_spots = [x[:] for x in board_matrix]
```
Comment (mihaitzzza) [inline game/modes.py:12]: Please use `copy.deepecopy()` discussed yesterday. It's more clear and easier to read. Also the variable name is not suggestive. What does actually `check_spots` mean ... isn't it a copy of the board?

File: game/modes.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+        new_board = set_choice(check_spots, choice, 'o')
```
Comment (mihaitzzza) [inline game/modes.py:13]: `set` express setting something, not returning something.

File: game/modes.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+        is_won, _ = check_status(new_board)
+        if is_won:
+            return True, choice
```
Comment (mihaitzzza) [inline game/modes.py:16]: I would init `is_won = False` (and `choice = None`) outside the `for` and just `break` the `for` if the game is won.

File: game/modes.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    return False, None
```
Comment (mihaitzzza) [inline game/modes.py:17]: `return is_won, choice`

File: game/modes.py [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+        found_cell, choice = _found_winning_cell(board_matrix, available_options)
+        if found_cell:
+            return choice
+        found_cell, choice = _found_blocking_cell(board_matrix, available_options)
+        if found_cell:
+            return choice
+        return random.choice(available_options)
+
+
+    corners = [x for x in available_options if x % 2 == 1 and x != 5]
```
Comment (mihaitzzza) [inline game/modes.py:57]: Don't compute the `corners` here since you're using them way lower. It's kind of confusing. You either init all the things in the top of the functions, either right when you need them.

File: game/modes.py [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+    winning, choice = _found_winning_cell(board_matrix, available_options)
+    if winning:
+        return choice
+    blocking_needed, choice = _found_blocking_cell(board_matrix, available_options)
+    if blocking_needed:
+        return choice
+
```
Comment (mihaitzzza) [inline game/modes.py:64]: Too much empty lines. One is enough.

File: game/modes.py [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+
+    if last_move % 2 == 1 and last_move != 5:
```
Comment (mihaitzzza) [inline game/modes.py:66]: I need you to explain me this `if`s :) You can send a message on the team's Slack channel.

File: game/modes.py [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+        last_move = choice
+
+        # Check if the game is won or not.
+        is_won, is_over = check_status(board_matrix)
+
+        if is_won:
+            winner = player_name
+        step += 1
+
+    print('Game Over!')
```
Comment (mihaitzzza) [inline game/modes.py:142]: Since you have a menu and this function is JUST running a game, the function who called this should account for winning states and stuff like this. So all the code from L142 to the end of the function should "go away".

File: game/player.py [hunk 26]
```diff
@@ -8,7 +8,10 @@ def get_current_player(step):
         name = 'Player 1'
         sign = 'x'
     else:
-        name = 'Player 2'
+        if not cpu:
+            name = 'Player 2'
+        else:
+            name = 'CPU'
         sign = 'o'
```
Comment (mihaitzzza) [inline game/player.py:15]: Please move this on top of the `if not cpu` to be more clear why you're doing `else` instead of directly `elif not cpu`.

File: main.py [hunk 27]
```diff
@@ -3,6 +3,7 @@
 import uuid
 import os.path
 
+
```
Comment (mihaitzzza) [inline main.py:6]: Why are you touching files that you should not? :)

**Review by mihaitzzza — CHANGES_REQUESTED** [review]

File: main.py (hunk 27 above)
Comment (Fiphee) [reply main.py:6]: It needed an extra space after imports. That's my post-excuse cel putin :))

File: game/menus.py [hunk 28]
```diff
@@ -31,37 +33,32 @@ def play_menu():
...
-        elif selection == '3':
-            on_play_menu = False
 
 def select_difficulty_menu():
     '''
     Displays CPU difficulties to choose from
     '''
-    on_difficulty_menu = True
-    while on_difficulty_menu:
+    selection = 0
```
Comment (mihaitzzza) [inline game/menus.py:47]: You either init with `None` or empty string. Stop changeing variable type.

File: game/modes.py [hunk 29]
```diff
@@ -74,11 +75,16 @@ def _computer_move(available_options, board_matrix, last_move, difficulty):
     if last_move % 2 == 0:
         if 5 in available_options:
             return 5
+    corners = [x for x in available_options if x % 2 == 1 and x != 5]
+    try:
+        return random.choice(corners)
+    except:
+        return random.choice(available_options)
```
Comment (mihaitzzza) [inline game/modes.py:82]: Is this because `corners` can be an empty array? If so, I found easier to use `return random.choice(corners or available_options)`.

**Review by mihaitzzza — CHANGES_REQUESTED** [review]

File: game/modes.py [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+
+
+logger = logging.getLogger(__name__)
+
+
+def _find_winning_cell(board_matrix, available_options, sign):
+    '''
+    Creates a copy of the current state of the board and tries all the available options with the sign argument
+    and checks if any choice wins, returns True or False if there is a winning choice, along with the the choice itself.
+    '''
```
Comment (PixLix1) [inline game/modes.py:16]: PEP: Triple double-quoted strings should be used for docstrings.

I would rephrase the comment as it is a bit too long and becomes hard to follow :)

```suggestion
    '''
    Creates a copy of the current state of the board and tries all the available options with the sign argument.
    Checks if any choice wins.
    Returns True or False if there is a winning choice, along with the the choice itself.
    '''
```

File: game/modes.py [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+from .player import get_current_player
+import logging
+import random
+import copy
+
+
+logger = logging.getLogger(__name__)
+
+
+def _find_winning_cell(board_matrix, available_options, sign):
```
Comment (PixLix1) [inline game/modes.py:12]: Is there a specific reason for using an underscore in the beginning of the function's name?
I think it should be something like:
```suggestion
def find_winning_cell(board_matrix, available_options, sign):
```

File: game/modes.py [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+        is_won, _ = check_status(new_board)
+        if is_won:
+            break
+    return is_won, choice
+
+
+def _check_cell(cell_number, board):
+    '''
+    returns the Data ('x', 'o' or None) from the given cell in the given board
+    '''
```
Comment (PixLix1) [inline game/modes.py:31]: 1. naming convention (unless there is another reason for using underscore at the beginning of the function's name)
2. PEP warning for triple double-quotes

File: game/modes.py [hunk 33]
```diff
@@ -0,0 +1,143 @@
...
+def _computer_move(available_options, board_matrix, last_move, difficulty):
+    '''
+    difficulties: 'easy' picks randomly from the available options
+    'medium' checks to for a winning move for itself, or blocks player from winning on next move. 
+    if there is no winning move next move, it chooses a random available option
+    'hard' checks for win or blocks player like the medium difficulty, but also strategically picks a move
+    if there is no winning or blocking move
+    
+    return:: choice
+    '''
```
Comment (PixLix1) [inline game/modes.py:46]: 1. naming convention (unless there is another reason for using underscore at the beginning of the function's name)
2. PEP warning for triple double-quotes

```suggestion
def _computer_move(available_options, board_matrix, last_move, difficulty):
    """
    difficulties: 'easy' picks randomly from the available options
    'medium' checks to for a winning move for itself, or blocks player from winning on next move. 
    if there is no winning move next move, it chooses a random available option
    'hard' checks for win or blocks player like the medium difficulty, but also strategically picks a move
    if there is no winning or blocking move
    
    return:: choice
    """
```

File: game/modes.py [hunk 34]
```diff
@@ -0,0 +1,143 @@
...
+                if choice % 2 == 0:
+                    return choice
+    if last_move % 2 == 0:
+        if 5 in available_options:
+            return 5
+    corners = [x for x in available_options if x % 2 == 1 and x != 5]
+    return random.choice(corners or available_options)
+
+
+def _player_move(board_matrix, available_options):
```
Comment (PixLix1) [inline game/modes.py:82]: naming convention
```suggestion
def player_move(board_matrix, available_options):
```

**Review by mihaitzzza — APPROVED** [review]
